aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -334,6 +334,258 @@ void cufft_clear_plan_cache_impl(int64_t device_index) {
+  c10::optional<CuFFTConfig> uncached_plan;
+  const CuFFTConfig * config = nullptr;
+
+  if (plan_cache.max_size() > 0) {
+    guard.lock();
+    if (plan_cache.max_size() > 0) {  // check again after acquiring the lock
+      config = &plan_cache.lookup(Params);
+    }
+  }
+
+  if (config == nullptr) {
+    uncached_plan.emplace(Params);
+    config = &uncached_plan.value();
+  }
+
+  auto & plan = config->plan();
+
+  // dims should always be chosen so a valid embedding is possible
+  TORCH_INTERNAL_ASSERT(!config->should_clone_input());
+
+  // prepare cufft for execution
+  CUFFT_CHECK(cufftSetStream(plan, at::cuda::getCurrentCUDAStream()));
+  auto workspace = at::empty({ config->workspace_size() }, at::device(at::kCUDA).dtype(at::kByte));
+  CUFFT_CHECK(cufftSetWorkArea(plan, workspace.data_ptr()));
+
+  // execute transform plan (potentially many times)
+  iter.serial_for_each([&](char** data, const int64_t* strides, int64_t size) {
+        // Execute one batch of FFTs
+        TORCH_INTERNAL_ASSERT(size == batch_size);
+        TORCH_INTERNAL_ASSERT(strides[0] == out_strides[0] * out_element_size);
+        TORCH_INTERNAL_ASSERT(strides[1] == in_strides[0] * in_element_size);
+        exec_cufft_plan(*config, data[1], data[0], forward);
+      }, {0, iter.numel()});
+}
+
+// Returns n s.t. the first n sorted_dims can be represented as a
+// cufft embedding. See NOTE [ cuFFT Embedded Strides ]
+int64_t num_embeddable_dims(IntArrayRef strides, IntArrayRef sorted_dims) {
+  const int64_t max_dims = std::min(
+      cufft_max_ndim, static_cast<int64_t>(sorted_dims.size()));
+  for (int64_t i = 1; i < max_dims; ++i) {
+    auto dim = sorted_dims[i];
+    auto dim_prev = sorted_dims[i - 1];
+    if (strides[dim] % strides[dim_prev] != 0) {
+      return i;
+    }
+  }
+  return max_dims;
+}
+
+// Calculates the normalization constant and applies it in-place to self
+// sizes is the sizes of a twosided tensor and dims are all transformed dims
+void _fft_apply_normalization(const Tensor& self, int64_t normalization, IntArrayRef sizes, IntArrayRef dims) {
+  auto norm = static_cast<fft_norm_mode>(normalization);
+  if (norm == fft_norm_mode::none) {
+    return;
+  }
+
+  int64_t signal_numel = 1;
+  for (auto dim : dims) {
+    signal_numel *= sizes[dim];
+  }
+  const double scale_denom = (norm == fft_norm_mode::by_root_n) ?
+    std::sqrt(signal_numel) : static_cast<double>(signal_numel);
+  self.div_(scale_denom);
+}
+
+}  // namespace (anonymous)
+
+Tensor _fft_r2c_cufft(const Tensor& self, IntArrayRef dim, int64_t normalization, bool onesided) {
+  TORCH_CHECK(self.is_floating_point());
+  auto input_sizes = self.sizes();
+  DimVector out_sizes(input_sizes.begin(), input_sizes.end());
+  auto last_dim = dim.back();
+  auto last_dim_halfsize = (input_sizes[last_dim]) / 2 + 1;
+  if (onesided) {
+    out_sizes[last_dim] = last_dim_halfsize;
+  }
+
+  auto input_strides = self.strides();
+  bool has_broadcasted_dims = false;
```

Comment:
Add a comment explaining the need to check for whether the tensor has a broadcasting dim.

Tensors can be overlapped in other ways, too. Would that also be an issue?

Reply:
>Tensors can be overlapped (two distinct indices referring to the same element) in other ways, too. Would that also be an issue?

This is purely working around the cuFFT error and not an actual correctness concern. I think cuFFT is being cautious because it's supposed to support inplace transforms with these plans, but in practice this PR never does inplace transforms in user-supplied tensors and so there are no risks I can think of.